tests/unit/action/test_bonsai_asset.py

```diff
@@ -0,0 +1,243 @@
+            name="test/asset",
+            version="1.2.3",
+        ))
+
+        assert result == dict(
+            name="test/asset",
+            state="present",
+            builds=[],
+        )
+
+    def test_bonsai_metadata_only(self, mocker):
+        bonsai_params = mocker.patch.object(bonsai, "get_asset_parameters")
+        bonsai_params.return_value = dict(
+            builds=[], labels=dict(a="b"), annotations=dict(c="d"),
+        )
+
+        result = bonsai_asset.ActionModule.build_asset_args(dict(
+            name="test/asset",
+            version="1.2.3",
+        ))
+
+        assert result == dict(
+            name="test/asset",
+            state="present",
+            builds=[],
+            annotations=dict(c="d"),
+            labels=dict(a="b"),
+        )
+
+    def test_user_metadata_only(self, mocker):
+        bonsai_params = mocker.patch.object(bonsai, "get_asset_parameters")
+        bonsai_params.return_value = dict(
+            builds=[1, 2, 3], labels=None, annotations=None,
+        )
+
+        result = bonsai_asset.ActionModule.build_asset_args(dict(
+            name="test/asset",
+            version="1.2.3",
+            labels=dict(my="label"),
+            annotations=dict(my="annotation"),
+        ))
+
+        assert result == dict(
+            name="test/asset",
+            state="present",
+            builds=[1, 2, 3],
+            annotations=dict(my="annotation"),
+            labels=dict(my="label"),
+        )
+
+    def test_mixed_metadata(self, mocker):
+        bonsai_params = mocker.patch.object(bonsai, "get_asset_parameters")
+        bonsai_params.return_value = dict(
+            builds=[], labels=dict(my="x", a="b"), annotations=dict(my="c"),
+        )
+
+        result = bonsai_asset.ActionModule.build_asset_args(dict(
+            name="test/asset",
+            version="1.2.3",
+            labels=dict(my="label"),
+            annotations=dict(my="annotation"),
+        ))
+
+        assert result == dict(
+            name="test/asset",
+            state="present",
+            builds=[],
+            annotations=dict(my="annotation"),
+            labels=dict(my="label", a="b"),
+        )
+
+    def test_rename(self, mocker):
+        bonsai_params = mocker.patch.object(bonsai, "get_asset_parameters")
+        bonsai_params.return_value = dict(
+            builds=[], labels=None, annotations=None,
+        )
+
+        result = bonsai_asset.ActionModule.build_asset_args(dict(
+            name="test/asset",
+            version="1.2.3",
+            rename="my-asset",
+        ))
+
+        assert result == dict(
+            name="my-asset",
+            state="present",
+            builds=[],
+        )
+
+    def test_fail_bonsai(self, mocker):
+        bonsai_params = mocker.patch.object(bonsai, "get_asset_parameters")
+        bonsai_params.side_effect = errors.BonsaiError("Bonsai bad")
+
+        with pytest.raises(errors.Error, match="Bonsai bad"):
```

Comment:
How about `with pytest.raises(errors.BonsaiError)`?

Reply:
I used this form because this is how the actual code is supposed to handle errors. Ansible stuff really does not care what erorr exactly was raised, all it needs to do is make sure it does not bubble up from the plugin and instead gracefully construct a return value.